ceremonies/src/benchmarking.rs

```diff
@@ -0,0 +1,377 @@
+where
+	<T as frame_system::Config>::BlockNumber: From<u64>,
+	<T as pallet_timestamp::Config>::Moment: From<u64>,
+{
+	let phase = encointer_scheduler::Pallet::<T>::current_phase();
+	let mut blocknr = frame_system::Pallet::<T>::block_number();
+	while phase == encointer_scheduler::Pallet::<T>::current_phase() {
+		blocknr += 1.into();
+		run_to_block::<T>(blocknr);
+	}
+}
+
+pub fn set_timestamp<T: Config>(t: T::Moment) {
+	let _ = pallet_timestamp::Pallet::<T>::set(RawOrigin::None.into(), t);
+}
+
+fn fake_last_attendance_and_get_proof<T: Config>(
+	prover: &sr25519::Pair,
+	cid: CommunityIdentifier,
+) -> ProofOfAttendance<T::Signature, T::AccountId>
+where
+	<T as frame_system::Config>::AccountId: From<sp_core::sr25519::Public>,
+	<T as Config>::Signature: From<sp_core::sr25519::Signature>,
+{
+	let cindex = encointer_scheduler::Pallet::<T>::current_ceremony_index();
+	let prover_account_id = account_id::<T>(&prover);
+	assert_ok!(encointer_balances::Pallet::<T>::issue(
+		cid,
+		&prover_account_id,
+		NominalIncome::from_num(1)
+	));
+	Pallet::<T>::fake_reputation(
+		(cid, cindex - 1),
+		&account_id::<T>(&prover),
+		Reputation::VerifiedUnlinked,
+	);
+	assert_eq!(
+		Pallet::<T>::participant_reputation((cid, cindex - 1), &prover_account_id),
+		Reputation::VerifiedUnlinked
+	);
+	let cindex = encointer_scheduler::Pallet::<T>::current_ceremony_index();
+	IssuedRewards::<T>::insert((cid, cindex - 1), 0, ());
+	let proof = create_proof_of_attendance::<T>(prover_account_id, cid, cindex - 1, &prover);
+	proof
+}
+
+pub fn last_event<T: Config>() -> Option<<T as frame_system::Config>::Event> {
+	let events = frame_system::Pallet::<T>::events();
+	if events.len() < 1 {
+		return None
+	}
+	let frame_system::EventRecord { event, .. } = &events[events.len() - 1];
+	Some(event.clone())
+}
+
+fn register_users<T: Config>(
+	cid: CommunityIdentifier,
+	num_newbies: u32,
+	num_reputables: u32,
+) -> Vec<sr25519::Pair>
+where
+	<T as frame_system::Config>::AccountId: From<sp_core::sr25519::Public>,
+	<T as Config>::Signature: From<sp_core::sr25519::Signature>,
+{
+	let mut users: Vec<sr25519::Pair> = vec![];
+	let mut proofs: Vec<ProofOfAttendance<T::Signature, T::AccountId>> = vec![];
+
+	let num_users_total = num_newbies + num_reputables;
+	// create users and fake reputation
+	for i in 0..num_users_total {
+		let p = sr25519::Pair::from_entropy(&[i as u8; 32], None).0;
+		users.push(p.clone());
+		if i < num_reputables {
+			proofs.push(fake_last_attendance_and_get_proof::<T>(&p.clone(), cid));
+		}
+	}
+
+	// register users
+	for i in 0..num_users_total {
+		let p = users[i as usize].clone();
+		if i < num_reputables {
+			assert_ok!(Pallet::<T>::register_participant(
+				RawOrigin::Signed(account_id::<T>(&(p.clone()))).into(),
+				cid,
+				Some(proofs[i as usize].clone())
+			));
+		} else {
+			assert_ok!(Pallet::<T>::register_participant(
+				RawOrigin::Signed(account_id::<T>(&(p.clone()))).into(),
+				cid,
+				None
+			));
+		}
+	}
```

Comment:
This can be a bit more rust ergonomic:

* & is a borrow; hence you don't need ownership here, so a clone does nothing. I think it is more ergonomic like this:

```
// iter() returns &Pair, into_iter() return Pair (owned value)
for (i, p) users.iter().enumerate() {
    let maybe_proof = None
    if i < num_reputables {
	 maybe_proof = Some(proofs[i].clone())
    }
        
    assert_ok!(Pallet::<T>::register_participant(
	              RawOrigin::Signed(account_id::<T>(p)).into(),
		      cid,
		      maybe_proof
	      ));
}
```

Reply:
sexy :)